### PR TITLE
CC-7635: implement DLQ reporter for malformed doc records

### PR DESCRIPTION
**Pull request by levzem**
## Problem
the connector currently can fail/ignore records that are malformed on the ES side. Reporting them using the new AK 2.6 reporter feature would be beneficial.

## Solution
Add the AK 2.6 reporter DLQ feature

##### Does this solution apply anywhere else?
- [ ] yes
- [x] no

## Test Strategy
modify unit tests to check whether the reporter is called

##### Testing done:
- [x] Unit tests
- [ ] Integration tests
- [ ] System tests
- [ ] Manual tests

## Release Plan

 
this is a new feature therefore pushing to `master`

Signed-off-by: Lev Zemlyanov <lev@confluent.io>

### CODE REVIEW

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java [hunk 1]
```diff
@@ -53,6 +54,7 @@
 
   private final Set<String> existingMappings;
   private final BehaviorOnMalformedDoc behaviorOnMalformedDoc;
+  private final ErrantRecordReporter reporter;
```
Comment (kkonstantine) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java:57]: this member variable (like the one above) is unused in this class. It's used in the `BulkProcessor` instead.

File: src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java [hunk 2]
```diff
@@ -449,10 +459,21 @@ private BulkResponse execute() throws Exception {
                   System.currentTimeMillis() - startTime
               );
             }
+            batch.forEach(r -> recordMap.remove(r));
```
Comment (kkonstantine) [inline src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:462]: how about? 
```suggestion
            recordMap.keySet().removeAll(batch);
```

File: src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java [hunk 3]
```diff
@@ -449,10 +459,21 @@ private BulkResponse execute() throws Exception {
...
+            for (R record : batch) {
+              SinkRecord original = recordMap.get(record);
+              if (original != null && reporter != null) {
+                reporter.report(
+                    original,
+                    new ConnectException("Bulk request failed: " + bulkRsp.getErrorInfo())
+                );
+              }
+            }
+            batch.forEach(r -> recordMap.remove(r));
```
Comment (kkonstantine) [inline src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:476]: same suggestion as above

File: src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java [hunk 4]
```diff
@@ -449,10 +459,21 @@ private BulkResponse execute() throws Exception {
...
               );
             }
+            batch.forEach(r -> recordMap.remove(r));
             return bulkRsp;
           } else if (responseContainsMalformedDocError(bulkRsp)) {
             retriable = bulkRsp.isRetriable();
             handleMalformedDoc(bulkRsp);
+            for (R record : batch) {
+              SinkRecord original = recordMap.get(record);
+              if (original != null && reporter != null) {
```
Comment (kkonstantine) [inline src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:469]: I read this and I'm thinking: should we make more things conditional to the `reporter` being not `null`?
At least this whole `for` loop does not need to run unless `reporter != null`, right?

File: src/test/java/io/confluent/connect/elasticsearch/bulk/BulkProcessorTest.java [hunk 5]
```diff
@@ -413,6 +428,8 @@ public void ignoreOrWarnOnMalformedDoc() throws InterruptedException {
         fail(e.getMessage());
       }
     }
+
+    verify(reporter, times(4)).report(any(), any());
```
Comment (kkonstantine) [inline src/test/java/io/confluent/connect/elasticsearch/bulk/BulkProcessorTest.java:432]: would it make sense to capture and assert the exact record argument?

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java [hunk 6]
```diff
@@ -139,6 +139,13 @@ public void start(Map<String, String> props, ElasticsearchClient client) {
           .setBehaviorOnNullValues(behaviorOnNullValues)
           .setBehaviorOnMalformedDoc(behaviorOnMalformedDoc);
 
+      try {
+        // may be null if DLQ not enabled
+        builder.setErrantRecordReporter(context.errantRecordReporter());
+      } catch (NoClassDefFoundError e) {
+        // Will occur in Connect runtimes earlier than 2.6
```
Comment (wicknicks) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java:146]: should we have a (warn?) message saying that errant record reporting will be disabled for this task?

File: src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java [hunk 7]
```diff
@@ -453,6 +462,18 @@ private BulkResponse execute() throws Exception {
           } else if (responseContainsMalformedDocError(bulkRsp)) {
             retriable = bulkRsp.isRetriable();
             handleMalformedDoc(bulkRsp);
+            if (reporter != null) {
+              for (R record : batch) {
+                SinkRecord original = recordMap.get(record);
+                if (original != null) {
+                  reporter.report(
+                      original,
+                      new ConnectException("Bulk request failed: " + bulkRsp.getErrorInfo())
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:471]: its protected - the only exception that sets the stack trace to not writable is `AbortException`

[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:471]: I found a way to supress the stack trace but that means creating a new exception and overriding the stack trace

File: src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java [hunk 8]
```diff
@@ -594,4 +617,16 @@ public String toString() {
       return name().toLowerCase(Locale.ROOT);
     }
   }
+
+  public static class NoStackTraceConnectException extends RuntimeException {
```
Comment (wicknicks) [inline src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:621]: maybe we can call this class `ReportingException`?. the exception doesn't have anything do with `ConnectException`.

File: src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java [hunk 9]
```diff
@@ -594,4 +617,16 @@ public String toString() {
...
+
+    public NoStackTraceConnectException(String message) {
+      super(message);
+    }
+
+    @Override
+    public synchronized Throwable fillInStackTrace() {
+      return this;
```
Comment (wicknicks) [inline src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java:629]: maybe add a javadoc explaining what we are doing here?

**Pull request merged**